### PR TITLE
feat: Add prefix and use colon separator in batch denoms (#392)

### PR DESCRIPTION
**Pull request by ruhatch**
## Description

This PR updates denominations for credit batches to have the form `C01-20190101-20201010-002`. This makes them human-readable, versionable, and easier to parse. They will also be compatible with the `x/bank` module using a prefix of `ueco:`.

This includes:
- Changing ClassIDs to be human readable, with credit type abbreviation and sequence number
- Storing class sequence numbers per credit type
- Changing batch denominations to use class id, dates, and batch sequence number
- Storing batch sequence numbers per credit class
- Updating CLI to use the new format and use spaces instead of commas to parse command line values

Closes: #392

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [x] added `!` to the type prefix if API or client breaking change
- [x] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] provided a link to the relevant issue or specification
- [x] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)
- [x] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [x] added a changelog entry to `CHANGELOG.md`
- [x] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [x] updated the relevant documentation or specification
- [x] reviewed "Files changed" and left comments if necessary
- [x] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: x/ecocredit/client/tx.go
Comment (aaronc) [inline x/ecocredit/client/tx.go:193]: Looks good. What do you think of hyphens for the class and batch separator? I.e.
```suggestion
            eg: 10eco:ABC-123,0.1eco:XYZ-456`,
```

I'm thinking of what these might look like when placed in a URI.

Comment (ruhatch) [reply x/ecocredit/client/tx.go:193]: I do find the multiple colons a bit odd.

> Also, for IID/DID compatibility we want to switch from : to / separators, i.e. 18AV53K:1b1vrHF.

I don't know what IID/DID is though, so not sure what the requirements for compatibility are. Is it just not having a `/`?

**Review by blushi — CHANGES_REQUESTED** [review]
Looks good overall, just a concern regarding limiting the max number of credit classes per credit type and batches per credit classes
https://github.com/regen-network/regen-ledger/pull/389 genesis import/export will need some updates after we merge this cc/ @aleem1314

File: x/ecocredit/denom.go
Comment (aaronc) [reply x/ecocredit/denom.go:16]: No! We can't do that. We just by default pad with 3 digits so we have `001` for `1`, but if we have `1000` then we just write that. Also if anything we should have 2 (or 3) digits for class and 3 (or 4) for batch. We expect many more batches than classes

**Review by aaronc — DISMISSED** [review]

File: x/ecocredit/denom.go
Comment (blushi) [inline x/ecocredit/denom.go:65]: could you add a test case for that?

**Review by blushi — APPROVED** [review]
tACK, pre-approving!

File: x/ecocredit/denom.go
Comment (technicallyty) [inline x/ecocredit/denom.go:16]:
```suggestion
	return fmt.Sprintf("%s%03d", creditType.Abbreviation, classSeqNo), nil
```

do we want 3 digits here? ref: https://github.com/regen-network/regen-ledger/pull/413#discussion_r685324116

**Review by technicallyty — APPROVED** [review]
pre-approve

File: x/ecocredit/denom.go
Comment (blushi) [reply x/ecocredit/denom.go:16]: using this formatting, if `classSeqNo` is 1000 for instance then it will just display 1000, if it's equal to 9, then 009 will be displayed, this works as expected

**Review by cyberbono3 — APPROVED** [review]
lgtm

File: x/ecocredit/denom.go
Comment (blushi) [reply x/ecocredit/denom.go:16]: oh right, I misread and thought that the `03d` formatting was already there
but aaron said that 2 digits (or 3) was ok for credit class and 3 (or 4 resp.) for credit batches, so let's just keep 2 and 3 which is the case atm

**Comment by blushi** [comment]
@aaronc should you have a final review on that or can we enable auto-merge?

**Comment by clevinson** [comment]
Adding auto merge, and we can always update if necessary when reviewing as part of the audit process.

**Pull request merged**
